staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval.go

```diff
@@ -0,0 +1,101 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cacher
+
+import "sync"
+
+// intervalTracker keeps track of watchCacheIntervals based on
+// the startIndex of each of these intervals. It is mainly used
+// for keeping track of intervals that include a specified startIndex
+// in its range - and if so, optionally invalidate those intervals.
+type intervalTracker struct {
+	watchCacheIntervals []*watchCacheInterval
+}
+
+func newIntervalTracker() *intervalTracker {
+	return &intervalTracker{}
+}
+
+// invalidateIntervals marks all watchCacheIntervals as invalid that
+// have refStartIndex as part of its interval.
+// TODO
+func (it *intervalTracker) invalidateIntervals(refStartIndex int) {
+}
+
+// TODO
+// this is where the heap related implementation goes.
+func (it *intervalTracker) trackInterval(interval *watchCacheInterval) {
+}
+
+// watchCacheInterval helps with keeping track of the interval of
+// items being considered while getting events since a particular
+// resourceVersion.
+type watchCacheInterval struct {
+	// startIndex denotes the starting point
+	// of interval being considered, the index
+	// value is the index in the actual watch
+	// cache, however this should be used modulo
+	// the capacity of the cache whenever being
+	// used to index the circular buffer.
+	startIndex int
+	// endIndex denotes the end of the interval
+	// being considered, the index value is the
+	// index in the actual watch cache, however
+	// this should be used modulo the capacity
+	// of the cache whenever being used to index
+	// the circular buffer.
+	endIndex int
+	// fillAll represents the case where the entire
+	// current state of the cache store needs to be
+	// returned, i.e. for cases where rv = 0.
+	fillAll bool
+	// indexer takes an index and returns the corresponding
+	// watchCacheEvent from an underlying watch cache.
+	indexer func(int) *watchCacheEvent
+	// lock effectively guards the valid field.
+	lock sync.Mutex
+	// valid indicates whether the cache interval
+	// being considered (b/w startIndex and endIndex)
+	// is still valid or not.
+	valid bool
+}
+
+// Next returns the next item in the cache interval provided the
+// cache interval is still valid.
+func (wci *watchCacheInterval) Next() (*watchCacheEvent, bool) {
+	wci.lock.Lock()
+	defer wci.lock.Unlock()
```

Comment:
Thinking a bit more about it - we will have to be a bit more sophisticated, because locking on every single item will be too expensive.

So basically the logic will be:

- we have a buffer [say of size 10 or 100]
```
Next() {
  if /* there is something in a buffer */ {
     return from buffer
  }
  wci.lock.Lock();
  defer wci.lock.Unlock()

  if !wci.valid {
    return nil, false
  }

  // copy items to the buffer (to fill this in)
  // when items copied to the buffer, we can already move the startIndex
  return the item from buffer
}
```